lib/adviser_csv_lookup.rb

```diff
@@ -0,0 +1,74 @@
+class AdviserCsvLookup
+  attr_reader :qualifications_lookup, :accreditations_lookup, :firms_lookup
+
+  def initialize(advisers)
+    @advisers = advisers
+  end
+
+  def build!
+    @qualifications_lookup = build_adviser_qualifications
+    @accreditations_lookup = build_adviser_accreditations
+    @firms_lookup          = build_adviser_firms
+  end
+
+  def certification_list
+    qualifications.values + accreditations.values
+  end
+
+  def each(&block)
+    @advisers.each do |adviser|
+      adviser_qualifications = qualifications_lookup[adviser.id] || []
+      adviser_accreditations = accreditations_lookup[adviser.id] || []
+      block.call(
+        adviser: adviser,
+        firm_name: firms_lookup[adviser.firm_id],
+        qualifications: adviser_qualifications,
+        accreditations: adviser_accreditations
+      )
+    end
+  end
+
+  private
+
+  def adviser_ids
+    @advisers.map(&:id)
+  end
+
+  def build_adviser_qualifications
+    certs = certification_query('advisers_qualifications', 'qualification')
+    aggregate_certificate_names(certs, type: 'qualification')
+  end
+
+  def build_adviser_accreditations
+    certs = certification_query('accreditations_advisers', 'accreditation')
+    aggregate_certificate_names(certs, type: 'accreditation')
+  end
+
+  def build_adviser_firms
+    Firm.pluck(:id, :registered_name).to_h
+  end
+
+  def certification_query(table, key)
+    ActiveRecord::Base.connection.execute("
+      SELECT adviser_id, #{key}_id
+      FROM #{table}
+      WHERE adviser_id IN (#{adviser_ids.join(',')})
+    ").to_a
+  end
+
+  def qualifications
+    @qualifications ||= Qualification.pluck(:id, :name).to_h
+  end
+
+  def accreditations
+    @accreditations ||= Accreditation.pluck(:id, :name).to_h
+  end
+
+  def aggregate_certificate_names(adviser_certifications, type:)
+    adviser_certifications.each_with_object({}) do |certification, h|
+      adviser_id = certification['adviser_id'].to_i
+      h[adviser_id] ||= []
+      h[adviser_id] << send(type.pluralize)[certification["#{type}_id"].to_i]
+    end
+  end
+end
```

Comment:
I will be honest that was kind of difficult to get my head around it. So I have questions and considerations about this class:

1) Is it not possible to make any sort of inner join from Adviser with accreditations and qualifications? This class sounds to me something wrong with the table relationship or with the mapping of the database, but I can be wrong. I am asking because then almost everything of this class could be removed if the answer is yes.
2) Do we need raw SQL to run a simple select query? Can Active Record be used?
3) There is the build! method with is called in the other class before the each method. This sound too much couple to do in other class. Can we not put inside the each method?
4) This class does not respect the single responsibility principle. There is a lot of going on on building the accreditation and qualification for the adviser plus getting the certification list plus aggregating the certificated names. I still need to think what could be a desired solution, but perhaps has one CertificationList class and this one.
5) I still didn't figured it out yet, but since we have a lot of private methods, perhaps there is some tests missing covering the behaviour. Like the aggregated method, where is it being tested?

Sorry, don't take personally, I understand the performance improvements, but I think we need to re look at this file. I would like to hear @margOnline opinions as well.

Reply:
Not a problem - answers below:

1. So `ActiveRecord`'s own cache was not fast enough for this to run in reasonable time. This meant that when we fetched an advisor and then attempted to fetch their qualifications/accreditations and their names, going across this many to many relation meant essentially a pretty bad bad n + 1.

We did attempt to write this as an inner join but, because of the join table in the middle, and the fact we wanted to actually output a list of all qualifications/accreditations filtered against the ones on an advisor the raw sql would have been incredibly complex and hard to follow.

2. The query is not simple because we're querying off of a join table (i.e. not a model). It would have been nice to be able to do:
```ruby
AdvisorQualifications.where(adviser_id: adviser_ids)
# and
AdvisorAccreditations.where(advisor_id: adviser_ids)
```
However this would involve creation of a new class, and essentially moving complication elsewhere.

3. Yep, I can move this 👍 

4. I can also have a look at splitting this apart into smaller classes too.

5. The classes is essentially tested on input/output. We're not testing the `each` method here - we made the lookup hashes for adviser qualifications, accreditations and firms available via `attr_readers` and test their specific structure and contents. This is the test that exercises the aggregate method. The private methods are the results of reducing duplication and refactoring, not adding additional untested behaviour.

Not a problem - we definitely would have liked to have stuck with `ActiveRecord`, but it was very clear from the outset that it wasn't able to deal with the many to many join table efficiently.